program/src/processor.rs

```diff
@@ -560,34 +575,34 @@ impl Processor {
 
     fn process_withdraw(accounts: &[AccountInfo], withdraw_lamports: u64) -> ProgramResult {
         let account_info_iter = &mut accounts.iter();
+
         let source_stake_account_info = next_account_info(account_info_iter)?;
-        let destination_stake_account_info = next_account_info(account_info_iter)?;
+        let destination_info = next_account_info(account_info_iter)?;
         let clock_info = next_account_info(account_info_iter)?;
-        let clock = &Clock::from_account_info(clock_info)?;
         let _stake_history_info = next_account_info(account_info_iter)?;
-        let stake_history = &StakeHistorySysvar(clock.epoch);
         let withdraw_authority_info = next_account_info(account_info_iter)?;
         let option_lockup_authority_info = next_account_info(account_info_iter).ok();
 
-        // this is somewhat subtle, but if the stake account is Uninitialized, you pass
-        // it twice and sign ie, Initialized or Stake, we use real withdraw
-        // authority. Uninitialized, stake account is its own authority
-        let (signers, custodian) = collect_signers(
-            &[withdraw_authority_info],
-            option_lockup_authority_info,
-            true,
-        )?;
+        let clock = &Clock::from_account_info(clock_info)?;
+        let stake_history = &StakeHistorySysvar(clock.epoch);
+
+        // this is somewhat subtle. for Initialized and Stake, there is a real authority
+        // but for Uninitialized, the source account is passed twice, and signed for
+        let (signers, custodian) =
+            collect_signers_checked(Some(withdraw_authority_info), option_lockup_authority_info)?;
```

Comment:
reproducing your comment from #7:
```
        // JC: do you actually need to pass it twice? I imagine in the current
        // version, you could just pass it once as a signer and it would Just Work,
        // but this falls in the same camp as the other similar comments: if
        // people are depending on that current weird behavior, we break 'em.
```

`Withdraw` has the dubious distinction of being the only stake program instruction (other than the relatively new `Move` instructions) that builds its own signers hashset and does it correctly. the `AuthorizeWithSeed` instructions do it wrong, and everything else uses the globbed hashset of all signers

you actually do need to pass the account twice for the `Uninitialized` case because the native program asserts there are at least five accounts and checks signer on the fifth. if you pass the same account twice with different `AccountMeta` privileges, it elevates the account to the highest, and works fine. if you pass the account once, it fails for not having enough accounts. if you pass a different account the second time, it fails to find the stake account address in the signers hashset and fails

Reply:
also i left the `if !signers.contains(source_stake_account_info.key) { ... }` block unchanged because this is already confusing enough